File: 9. Palindrome Number/solution.java
// 按位取x，逆转成int，比较  但是不能兼容溢出的情况，如果逆转后溢出了就会出错。
class Solution {
    public boolean isPalindrome(int x) {
        if (x < 0) {
            return false;
        }
        int origin = x;
        int reversed = 0;
        
        while (x != 0) {
            reversed = reversed * 10 + x % 10;
            x /= 10;
        }
        
        return reversed == origin;
    }
}

//按位取，但是保存每一位，然后遍历比较，可以兼容溢出的情况，因为没有重新生成int
class Solution {
    public boolean isPalindrome(int x) {
        if (x < 0)
            return false;
        List<Integer> list = new ArrayList<>();
        while (x > 0) {
            list.add(x % 10);
            x /= 10;
        }
        int i = 0, j = list.size() - 1;
        while (i < j) {
            if (list.get(i) != list.get(j))
                return false;
            i++;
            j--;
        }
        return true;
    }
}

File: 9. Palindrome Number/solution.py
# 切片，x[::-1]倒序生成新列表，跟原列表比较
class Solution:
    def isPalindrome(self, x):
        """
        :type x: int
        :rtype: bool
        """
        if(x >= 0):
            return str(x) == str(x)[::-1]
        else:
            return False

# 将数字逆过来，跟原数字比较
class Solution:
    def isPalindrome(self, x):
        """
        :type x: int
        :rtype: bool
        """
        r = 0
        temp = x
        if x < 0:
            return False
        while temp > 0:
            r = r * 10 + temp % 10
            temp //= 10
        if r == x:
            return True
        else:
            return False
